src/wasm-type-printing.h

```diff
@@ -93,6 +94,33 @@ struct IndexedTypeNameGenerator
   }
 };
 
+// Prints heap types stored in a module, falling back to the given
+// FallbackGenerator if the module does not have a name for type type.
+template<typename FallbackGenerator = DefaultTypeNameGenerator>
+struct ModuleTypeNameGenerator
+  : TypeNameGeneratorBase<ModuleTypeNameGenerator<FallbackGenerator>> {
+  const Module& wasm;
+  DefaultTypeNameGenerator defaultGenerator;
+  FallbackGenerator& fallback;
+
+  ModuleTypeNameGenerator(const Module& wasm, FallbackGenerator& fallback)
+    : wasm(wasm), fallback(fallback) {}
+
+  // TODO: Use C++20 `requires` to clean this up.
+  template<class T = FallbackGenerator>
+  ModuleTypeNameGenerator(
+    const Module& wasm,
+    std::enable_if_t<std::is_same_v<T, DefaultTypeNameGenerator>>* = nullptr)
+    : ModuleTypeNameGenerator(wasm, defaultGenerator) {}
+
+  TypeNames getNames(HeapType type) {
+    if (auto it = wasm.typeNames.find(type); it != wasm.typeNames.end()) {
+      return it->second;
+    }
+    return fallback.getNames(type);
```

Comment:
Not new in this PR, but `getNames` should maybe return a reference and not a copy..?

Reply:
I don't think that's possible, since in general the generators may be creating new names on the fly.